keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -1,113 +1,560 @@
+  //
+  // * Every time it tries to validate a JWT carrying an unknown key ID, and
+  //
+  // * Periodically, so it can serve reasonably-up-to-date keys from the OIDC
+  //   JWKs endpoint.
+  rpc FetchKeys(FetchKeysRequest) returns (FetchKeysResponse) {}
+  
+  // Metadata is meant to be called once on startup.
+  // Enables sharing metadata with kube-apiserver (eg: the max token lifetime that signer supports)
+  rpc Metadata(MetadataRequest) returns (MetadataResponse) {}
+}
 
-service KeyService {
-  // Sign an incoming payload
-  rpc SignPayload(SignPayloadRequest) returns (SignPayloadResponse) {}
-  // List all active public keys
-  rpc ListPublicKeys(ListPublicKeysRequest) returns (ListPublicKeysResponse) {}
+message SignJWTRequest {
+  // URL-safe base64 wrapped payload to be signed.
+  // Exactly as it appears in the second segment of the JWT
+  bytes claims = 1;
 }
 
-message SignPayloadRequest {
-  // payload is the content to be signed. JWT headers must be included by the caller
-  bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
+message SignJWTResponse {
+  // header must contain alg, kid, typ claims.
+  // type must be “JWT”.
+  // kid must be non-empty.
+  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).
+  // header cannot have any additional data that kube-apiserver does not recognize.
+  // Already wrapped in URL-safe base64, exactly as it appears in the first segment of the JWT.
+  string header = 1;
+
+  // The signature for the JWT.
+  // Already wrapped in URL-safe base64, exactly as it appears in the final segment of the JWT.
+  string signature = 2;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message FetchKeysRequest {}
+
+message FetchKeysResponse {
+  repeated Key keys = 1;
+
+  // The timestamp when this data was pulled from the authoritative source of
+  // truth for verification keys.
+  // kube-apiserver can export this from metrics, to enable end-to-end SLOs.
+  google.protobuf.Timestamp data_timestamp = 2;
+
+  // The timestamp for when it's appropriate to refetch verification keys to pick changes if any.
+  // Will be treated as a hint to compute cache max age (`suggested_next_fetch_at.Sub(time.Now())/time.Second`).
+  // max age will have a minimum floor of 10s (for both, kube-apiserver and OIDC provider) to avoid hot loop. 
+  google.protobuf.Timestamp suggested_next_fetch_at = 3;
 }
 
+message Key {
+  // A unique identifier for this key.
+  string key_id = 1;
 
-message PublicKey {
-  // public_key is a PEM encoded public key
-  bytes public_key = 1;
-  // certificate is a concatenated list of PEM encoded x509 certificates
-  bytes certificates = 2;
-  // key_id is the key's ID
-  string key_id = 3;
-  // algorithm states the algorithm the key uses
-  string algorithm = 4;
+  // The public key, PKIX-serialized.
+  bytes key = 2;
 }
```

Comment:
Clusters that 1) move to external signers, 2) keep supporting legacy tokens (either via a static signer or a custom external signer controller), and 3) use a distinct signing key for legacy tokens _should_ be able to keep that legacy token signing key from showing up in oidc discovery (since legacy tokens aren't oidc-shaped and wouldn't be verifiable by oidc-discovery-doc consumers anyway)

To enable that, add an `exclude_from_oidc_discovery bool` to PublicKey to allow the external signer to return a key that can be used for verification but excluded from oidc discovery. Document it should only be set for verifying keys which are not used to issue bound tokens.

Reply:
Done

also called out the Legacy token scenario is a section ("Support for Legacy Tokens") under "Notes/Constraints/Caveats" section